### ISSUE TITLE
Recursive types not supported in Go codegen

### ISSUE THREAD

**Issue by mikhailshilkov** [pulumi/pulumi#7622]
Unfortunately, we haven't been building the Go SDK of Google Native yet, so I don't know when this got broken (potentially, long time ago). Anyway, filing an issue to investigate. All other SDKs are fine, so I assume there's an issue with Go code generation specifically.

```
Error: 	/home/runner/work/pulumi-google-native/pulumi-google-native/sdk/go/google/bigquery/v2/pulumiTypes.go:16327:6: QueryParameterTypeResponse refers to
Error: 	/home/runner/work/pulumi-google-native/pulumi-google-native/sdk/go/google/bigquery/v2/pulumiTypes.go:16327:6: QueryParameterTypeResponse
go/google/bigquery/v2/pulumiTypes.go:19181:6: invalid recursive type StandardSqlDataTypeResponse
Error: 	/home/runner/work/pulumi-google-native/pulumi-google-native/sdk/go/google/bigquery/v2/pulumiTypes.go:19181:6: StandardSqlDataTypeResponse refers to
Error: 	/home/runner/work/pulumi-google-native/pulumi-google-native/sdk/go/google/bigquery/v2/pulumiTypes.go:19181:6: StandardSqlDataTypeResponse
# github.com/pulumi/pulumi-google-native/sdk/go/google/dialogflow/v3
go/google/dialogflow/v3/pulumiTypes.go:6603:6: GoogleCloudDialogflowCxV3IntentInput redeclared in this block
	previous declaration at go/google/dialogflow/v3/pulumiTypes.go:6355:6
# github.com/pulumi/pulumi-google-native/sdk/go/google/dialogflow/v3beta1
go/google/dialogflow/v3beta1/pulumiTypes.go:6627:6: GoogleCloudDialogflowCxV3beta1IntentInput redeclared in this block
	previous declaration at go/google/dialogflow/v3beta1/pulumiTypes.go:6379:6
# github.com/pulumi/pulumi-google-native/sdk/go/google/transcoder/v1beta1
go/google/transcoder/v1beta1/pulumiTypes.go:1784:69: cannot use v.Inputs (type []AudioChannelInput) as type []AudioChannelInput in return argument
go/google/transcoder/v1beta1/pulumiTypes.go:1808:6: AudioChannelInput redeclared in this block
	previous declaration at go/google/transcoder/v1beta1/pulumiTypes.go:1717:6
```

See https://github.com/pulumi/pulumi-google-native/pull/165/checks?check_run_id=3143114570

**Comment by mikhailshilkov** [comment]
A similar failure now happens in AWS Native:

```
# github.com/pulumi/pulumi-aws-native/sdk/go/aws/mediaconnect
go/aws/mediaconnect/flow.go:119:43: impossible type assertion:
	FlowOutput does not implement pulumi.Output (missing ApplyT method)
go/aws/mediaconnect/flowOutput.go:15:6: FlowOutput redeclared in this block
	previous declaration at go/aws/mediaconnect/flow.go:122:6
```

**Comment by emiliza** [comment]
Should be fixed by https://github.com/pulumi/pulumi/issues/7595

**Comment by EvanBoyle** [comment]
So there are a couple of errors here that persist even after #7595. 

The schema does indeed include recursive types which I don't believe our supported currently by our go codegen:

```go
type QueryParameterTypeResponse struct {
	// [Optional] The type of the array's elements, if this is an array.
	ArrayType QueryParameterTypeResponse `pulumi:"arrayType"`
	// [Optional] The types of the fields of this struct, in order, if this is a struct.
	StructTypes []QueryParameterTypeStructTypesItemResponse `pulumi:"structTypes"`
	// [Required] The top level type of this field.
	Type string `pulumi:"type"`
}
```

would need to be 
```go
type QueryParameterTypeResponse struct {
	// [Optional] The type of the array's elements, if this is an array.
	ArrayType *QueryParameterTypeResponse `pulumi:"arrayType"`
	// [Optional] The types of the fields of this struct, in order, if this is a struct.
	StructTypes []QueryParameterTypeStructTypesItemResponse `pulumi:"structTypes"`
	// [Required] The top level type of this field.
	Type string `pulumi:"type"`
}
```

Is there a way to remodel the types, or should we file a separate issue to support recursive types?

Second, there are a few redeclared type errors. This is due to collision with another reserved prefix `Input` for example a type of the name `GoogleCloudDialogflowCxV3beta1IntentInput` generates an input interface also named `GoogleCloudDialogflowCxV3beta1IntentInput` (also `AudioChannelInput`). 

We have seen this issue previously with other reserved suffixes (I believe `Args`?). We have two courses of action:

1. remap the types in the schema
2. add suffix remapping for types ending in `Input`.

**Comment by t0yv0** [comment]
Hm.. I don't know what's the immediate best course of action to unblock. For later, I think both issues should be bugs against codegen, codegen should take ownership of either supporting recursive types or erroring intelligently, and it should take ownership of detecting name collisions (and possibly giving easy options to resolve them?).

**Comment by mikhailshilkov** [comment]
I'm fairly sure azure-native has some recursive types too, but maybe in a meaningfully different way.

I agree with @t0yv0 here... I'm not even sure how to remodel those recursive types, so, ideally, the codegen should take care of those, as it apparently does for other languages.

**Comment by t0yv0** [comment]
@EvanBoyle can you narrow down the repro or d'you want me to try? Thanks!

**Comment by EvanBoyle** [comment]
The recursive type issue still hasn't been addressed.
